### PR TITLE
Defer handler reflection

### PR DESCRIPTION
**Pull request by SimonCropp**
no point doing the reflection on every send. only do it when _requestHandlers does not already contain a cached instance

### CODE REVIEW

File: src/MediatR/Mediator.cs [hunk 1]
```diff
@@ -48,19 +48,22 @@ public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, Cancellation
...
-            {
-                throw new ArgumentException($"{nameof(request)} does not implement ${nameof(IRequest)}");
-            }
-
-            var responseType = requestInterfaceType!.GetGenericArguments()[0];
             var handler = _requestHandlers.GetOrAdd(requestType,
-                t => Activator.CreateInstance(typeof(RequestHandlerWrapperImpl<,>).MakeGenericType(requestType, responseType)));
+                t =>
+                {
+                    var requestInterfaceType = requestType
```
Comment (jbogard) [inline src/MediatR/Mediator.cs:54]: Should this use `t` to avoid allocations?

Comment (SimonCropp) [reply src/MediatR/Mediator.cs]: good catch. fixed

File: src/MediatR/Mediator.cs [hunk 2]
```diff
@@ -48,19 +48,22 @@ public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, Cancellation
...
+                requestTypeKey =>
+                {
+                    var requestInterfaceType = requestTypeKey
+                        .GetInterfaces()
+                        .FirstOrDefault(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IRequest<>));
+                    var isValidRequest = requestInterfaceType != null;
+
+                    if (!isValidRequest)
+                    {
+                        throw new ArgumentException($"{nameof(request)} does not implement ${nameof(IRequest)}");
```
Comment (no1melman) [inline src/MediatR/Mediator.cs:61]: Moving this here - will it cause any issues?

Comment (jbogard) [reply src/MediatR/Mediator.cs:61]: Stack trace would be different?

Comment (no1melman) [reply src/MediatR/Mediator.cs:61]: Yeah that would be part of it. Which may lead it to be hidden?

Comment (jbogard) [reply src/MediatR/Mediator.cs:61]: Hidden by what? Are you suggesting the exception will get swallowed somehow?

Comment (SimonCropp) [reply src/MediatR/Mediator.cs:61]: i added a test for that scenario

File: src/MediatR/Mediator.cs [hunk 3]
```diff
@@ -58,7 +58,7 @@ public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, Cancellation
 
                     if (!isValidRequest)
                     {
-                        throw new ArgumentException($"{nameof(request)} does not implement ${nameof(IRequest)}");
+                        throw new ArgumentException($"{requestType.Name} does not implement {nameof(IRequest)}");
```
Comment (SimonCropp) [inline src/MediatR/Mediator.cs:61]: i also found a bug with the exception message

Comment (jbogard) [reply src/MediatR/Mediator.cs:61]: Should we add the argument name here? `nameof(requestType)`?

Comment (SimonCropp) [reply src/MediatR/Mediator.cs]: done

**Pull request merged**